File: examples/echo/src/main/scala/com/twitter/example/EchoLoadTest.scala
package com.twitter.example

import org.apache.thrift.protocol.TBinaryProtocol

import com.twitter.iago.processor.ThriftRecordProcessor
import com.twitter.iago.server.{ParrotRequest, ParrotService}
import com.twitter.logging.Logger

import thrift.EchoService

class EchoLoadTest(parrotService: ParrotService[ParrotRequest, Array[Byte]])
    extends ThriftRecordProcessor(parrotService) {
  val client = new EchoService.FinagledClient(service, new TBinaryProtocol.Factory())
  val log = Logger.get(getClass)

  override def processLine(line: String) {
    client.echo(line) respond { rep =>
      if (rep == "hello") {
        client.echo("OMIGOD IT'S TALKING TO US")
      }
      log.info("response: " + rep)
    }
  }
}

File: examples/echo/src/main/java/com/twitter/jexample/EchoLoadTest.java
package com.twitter.jexample;

import com.twitter.example.thrift.EchoService;
import com.twitter.finagle.stats.NullStatsReceiver;
import com.twitter.iago.processor.ThriftLoadTest;
import com.twitter.iago.server.ParrotRequest;
import com.twitter.iago.server.ParrotService;
import com.twitter.util.Future;
import com.twitter.util.FutureEventListener;

import org.apache.thrift.protocol.TBinaryProtocol;

import java.util.List;

public class EchoLoadTest extends ThriftLoadTest {
  EchoService.FinagledClient client = null;

  public EchoLoadTest(ParrotService<ParrotRequest, byte[]> parrotService) {
    super(parrotService);
    client = new EchoService.FinagledClient(
        service(),
        new TBinaryProtocol.Factory(),
        "EchoService",
        new NullStatsReceiver());
  }

  public void processLine(String line) {
    Future<String> future = client.echo(line);
    future.addEventListener(new FutureEventListener<String>() {
      public void onSuccess(String msg) {
        System.out.println("response: " + msg);
      }

      public void onFailure(Throwable cause) {
        System.out.println("Error: " + cause);
      }
    });
  }
}
